testapp/src/main/java/com/google/android/agera/testapp/RowHandler.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.android.agera.testapp;
+
+import static com.google.android.agera.Preconditions.checkNotNull;
+import static com.google.android.agera.Repositories.mutableRepository;
+import static com.google.android.agera.rvadapter.RepositoryAdapter.repositoryAdapter;
+
+import com.google.android.agera.Binder;
+import com.google.android.agera.Function;
+import com.google.android.agera.MutableRepository;
+import com.google.android.agera.Receiver;
+import com.google.android.agera.rvadapter.RepositoryAdapter;
+import com.google.android.agera.rvadapter.RepositoryPresenter;
+
+import android.os.Parcelable;
+import android.support.annotation.NonNull;
+import android.support.v7.widget.RecyclerView;
+import android.support.v7.widget.RecyclerView.Adapter;
+import android.support.v7.widget.RecyclerView.LayoutManager;
+import android.support.v7.widget.RecyclerView.OnScrollListener;
+import android.support.v7.widget.RecyclerView.RecycledViewPool;
+import android.view.View;
+
+import java.util.HashSet;
+import java.util.IdentityHashMap;
+import java.util.Map;
+import java.util.Set;
+
+final class RowHandler<TRow, TRowItems> extends OnScrollListener
+    implements Binder<TRow, View>, Receiver<View> {
+  @NonNull
+  private final Set<RepositoryAdapter> startedAdapters;
+  @NonNull
+  private final Map<Long, Parcelable> itemRowStates;
+  @NonNull
+  private final Map<Adapter, MutableRepository<TRowItems>> adapterRepositories;
```

Comment:
Not too sure if you need to effectively keep all adapters and horizontal repositories in the memory. You can probably remove the recycled ones as they are recycled.